File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.cpp
class Solution {
public:
    int myAtoi(string s) {
        s.erase(0, s.find_first_not_of(' '));

        if (s.empty()) return 0;

        bool isneg = (s[0] == '-');
        if (s[0] == '-' || s[0] == '+') {
            s = s.substr(1);
        }

        string ans = "";
        for (char c : s) {
            if (!isdigit(c)) break;
            ans += c;
        }

        if (ans.empty()) return 0;

        try {
            long long num = stoll(ans);
            if (isneg) num = -num;

            if (num < INT_MIN) return INT_MIN;
            if (num > INT_MAX) return INT_MAX;
            return static_cast<int>(num);
        } catch (...) {

            return isneg ? INT_MIN : INT_MAX;
        }
    }
};

File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.java
class Solution {
    static int mx = 2147483647, mn = -2147483648;
    static int helper(String s, int i, long res, int sign){
        if(i >= s.length() || !Character.isDigit(s.charAt(i)))
            return (int)(sign*res);
        res = res*10 + (s.charAt(i) - '0');

        if(sign*res >= mx) return mx;
        if(sign*res <= mn) return mn;

        return helper(s, i+1, res, sign);
    }
    public int myAtoi(String s) {
        int i=0;
        while(i < s.length() && (s.charAt(i) == ' ')) i++;
        
        int sign = 1;
        if(i < s.length() && (s.charAt(i) == '-' || s.charAt(i) == '+')){
            sign = (s.charAt(i) == '+')? 1:-1;
            i++;
        }

        return helper(s, i, 0, sign);
    }
}